lib/membrane_ffmpeg_video_filter/text_overlay.ex

```diff
@@ -90,50 +123,98 @@ defmodule Membrane.FFmpeg.VideoFilter.TextOverlay do
+    {{:ok, [buffer: {:output, buffer}]}, state}
+  end
+
+  # no text left to render
+  defp apply_filter_if_needed(buffer, _ctx, %{text_stream: []} = state) do
+    {buffer, state}
+  end
+
+  defp apply_filter_if_needed(
+         %{metadata: metadata, payload: payload} = buffer,
+         ctx,
+         %{native_state: native_state, text_stream: [{interval, _text} | streams]} = state
+       ) do
+    cond do
+      frame_before_interval?(metadata, interval) ->
+        {buffer, state}
+
+      frame_after_interval?(metadata, interval) ->
+        state = %{state | text_stream: streams}
+        state = init_new_filter_if_needed(ctx.pads.input.caps, state)
+        apply_filter_if_needed(buffer, ctx, state)
+
+      frame_in_interval?(metadata, interval) ->
+        case Native.apply_filter(payload, native_state) do
+          {:ok, frame} ->
+            {%{buffer | payload: frame}, state}
+
+          {:error, reason} ->
+            raise inspect(reason)
+        end
+    end
+  end
+
+  defp init_new_filter_if_needed(_caps, %{text_stream: []} = state), do: state
+
+  defp init_new_filter_if_needed(caps, %{text_stream: [stream | _streams]} = state) do
+    {_interval, text} = stream
+
     case Native.create(
-           state.text,
-           width,
-           height,
-           format,
+           text,
+           caps.width,
+           caps.height,
+           caps.format,
            state.fontsize,
            state.box?,
            state.boxcolor,
            state.border?,
            state.fontcolor,
            fontfile_to_native_format(state.fontfile),
-           state.vertical_align,
-           state.horizontal_align
+           state.horizontal_align,
+           state.vertical_align
          ) do
       {:ok, native_state} ->
-        state = %{state | native_state: native_state}
-        {{:ok, caps: {:output, caps}, redemand: :output}, state}
+        %{state | native_state: native_state}
 
       {:error, reason} ->
         raise inspect(reason)
     end
   end
 
-  @impl true
-  def handle_process(
-        :input,
-        %Buffer{payload: payload} = buffer,
-        _ctx,
-        %{native_state: native_state} = state
-      ) do
-    case Native.apply_filter(payload, native_state) do
-      {:ok, frame} ->
-        buffer = [buffer: {:output, %{buffer | payload: frame}}]
-        {{:ok, buffer}, state}
+  defp frame_before_interval?(metadata, {from, _to}) do
+    pts = get_pts(metadata)
+    pts < from
+  end
 
-      {:error, reason} ->
-        raise inspect(reason)
-    end
+  defp frame_after_interval?(_metadata, {_from, :infinity}), do: false
+
+  defp frame_after_interval?(metadata, {_from, to}) do
+    pts = get_pts(metadata)
+    pts >= to
+  end
+
+  defp frame_in_interval?(metadata, {from, :infinity}) do
+    pts = get_pts(metadata)
+    pts >= from
+  end
+
+  defp frame_in_interval?(metadata, {from, to}) do
+    pts = get_pts(metadata)
+    pts < to and pts >= from
```

Comment:
If PTS can be ratio, you should use comparison functions from Ratio

Reply:
PTS is returned as integer here